presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -1883,6 +1906,91 @@ public ConnectorPushdownFilterResult pushdownFilter(ConnectorSession session, Co
+
+        public ConstraintEvaluator(RowExpressionService evaluator, ConnectorSession session, Map<String, ColumnHandle> assignments, RowExpression expression)
+        {
+            this.assignments = assignments;
+            this.evaluator = evaluator;
+            this.session = session;
+            this.expression = expression;
+
+            arguments = ImmutableSet.copyOf(extractAll(expression)).stream()
+                    .map(VariableReferenceExpression::getName)
+                    .map(assignments::get)
+                    .collect(toImmutableSet());
+        }
+
+        private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
+        {
+            if (intersection(bindings.keySet(), arguments).isEmpty()) {
+                return true;
+            }
+            LookupVariableResolver inputs = new LookupVariableResolver(assignments, bindings);
+
+            // Skip pruning if evaluation fails in a recoverable way. Failing here can cause
+            // spurious query failures for partitions that would otherwise be filtered out.
+            Object optimized = null;
+            try {
+                optimized = evaluator.optimizeRowExpression(session, expression, inputs, OPTIMIZED);
+            }
+            catch (PrestoException e) {
+                propagateIfUnhandled(e);
+            }
+
+            // If any conjuncts evaluate to FALSE or null, then the whole predicate will never be true and so the partition should be pruned
+            return !Boolean.FALSE.equals(optimized) && optimized != null && (!(optimized instanceof ConstantExpression) || !((ConstantExpression) optimized).isNull());
+        }
+
+        private static void propagateIfUnhandled(PrestoException e)
+                throws PrestoException
+        {
+            int errorCode = e.getErrorCode().getCode();
+            if (errorCode == DIVISION_BY_ZERO.toErrorCode().getCode()
+                    || errorCode == INVALID_CAST_ARGUMENT.toErrorCode().getCode()
+                    || errorCode == INVALID_FUNCTION_ARGUMENT.toErrorCode().getCode()
+                    || errorCode == NUMERIC_VALUE_OUT_OF_RANGE.toErrorCode().getCode()) {
+                return;
+            }
+
+            throw e;
+        }
+    }
+
+    private static class LookupVariableResolver
+            implements Function<VariableReferenceExpression, Object>
+    {
+        private final Map<String, ColumnHandle> assignments;
+        private final Map<ColumnHandle, NullableValue> bindings;
+
+        public LookupVariableResolver(
+                Map<String, ColumnHandle> assignments,
+                Map<ColumnHandle, NullableValue> bindings)
+        {
+            this.assignments = requireNonNull(assignments, "assignments is null");
+            this.bindings = ImmutableMap.copyOf(requireNonNull(bindings, "bindings is null"));
+        }
+
+        @Override
+        public Object apply(VariableReferenceExpression variable)
```

Comment:
@wenleix Sure. Like so?

```
            Function<VariableReferenceExpression, Object> variableResolver = variable -> {
                ColumnHandle column = assignments.get(variable.getName());
                checkArgument(column != null, "Missing column assignment for %s", variable);

                if (!bindings.containsKey(column)) {
                    return variable;
                }

                return bindings.get(column).getValue();
            };

            // Skip pruning if evaluation fails in a recoverable way. Failing here can cause
            // spurious query failures for partitions that would otherwise be filtered out.
            Object optimized = null;
            try {
                optimized = evaluator.optimizeRowExpression(session, expression, variableResolver, OPTIMIZED);
            }
            catch (PrestoException e) {
                propagateIfUnhandled(e);
            }
```